File: chapter 09/recipe 9.3/DniScVoting/DniScAutoVoteObj.cs
using System;
using System.EnterpriseServices;
using System.Runtime.InteropServices;

using TransactionLogging;

namespace DniScVoting
{
	[Transaction(TransactionOption.Required)]
	[ClassInterface(ClassInterfaceType.None)]
	public class DniScAutoVoteObj
		: ServicedComponent, ITranMethods
	{
		[AutoComplete]
		public void PerformWork(RequestedResult request)
		{
			//log the transaction
			TransactionLogger log
				= new TransactionLogger(this);

			//determine what kind of result was requested
			switch (request)
			{
				case RequestedResult.VoteToAbort:
					ContextUtil.MyTransactionVote
						= TransactionVote.Abort;
					break;
				case RequestedResult.ThrowException:
					throw new ApplicationException(
						"Transaction should be aborted");
				default:
					break;
			}
		}
	}
}

File: chapter 09/recipe 9.3/DniScVotingVB/DniScAutoVoteObj.vb
Imports System.Runtime.InteropServices
Imports System.EnterpriseServices

<Transaction(TransactionOption.Required)> _
<ClassInterface(ClassInterfaceType.None)> _
Public Class DniScAutoVoteObj
    Inherits ServicedComponent
    Implements ITranMethods

    <AutoComplete()> _
    Public Sub PerformWork(ByVal request As RequestedResult) _
            Implements ITranMethods.PerformWork
        'log the transaction
        Dim log As TransactionLogging.TransactionLogger _
            = New TransactionLogging.TransactionLogger(Me)

        'determine what kind of result was requested
        Select Case request
            Case RequestedResult.VoteToAbort
                ContextUtil.MyTransactionVote _
                    = TransactionVote.Abort
            Case RequestedResult.ThrowException
                Throw New ApplicationException( _
                    "Transaction should be aborted")
        End Select
    End Sub
End Class
